### PR TITLE
fix(setup-window-mock): reset in `afterEach` hook

### PR DESCRIPTION
**Pull request by buschtoens**
Runs the `reset` in `afterEach` to clean up the `window` mock after every test. This assumes that all preceding tests did not leak state and the test module starts with a pristine `window` mock.

This PR _could_ potentially make your tests fail. If so, you somewhere have a leaky preceding test.

---

I have some tests which indirectly depend on the default state of the global `window` mock (all imported via `ember-window-mock`) and they started failing after I mutated that state in an unrelated test.

While I _technically could_ and _probably should_ (?) call `setupWindowMock` in the failing tests as well, I think it would be easiest and safest, if every test module would clean up after itself and reset the global `window` mock back to its default state.

### CODE REVIEW

File: addon-test-support/-private/setup-window-mock.js [hunk 1]
```diff
@@ -7,4 +7,5 @@ import { reset } from 'ember-window-mock';
 //
 export default function setupWindowMock(hooks = self) {
   hooks.beforeEach(reset);
+  hooks.after(reset);
```
[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: I could swear that they wear, last time I checked. 😂

Pardon me. I'll update the PR.

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: name | description
-- | --
before (function) | Runs before the first test.
beforeEach (function) | Runs before each test.
afterEach (function) | Runs after each test.
after (function) | Runs after the last test.

Source: https://api.qunitjs.com/QUnit/module

Seems like the `before` and `after` hooks get underway... I'll investigate.

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: Ah, so the reason this was failing is that the `after` spy was missing:
https://github.com/kaliber5/ember-window-mock/pull/79/commits/f1407aab10bf5a68e143c8a24508f073f7c07f34#diff-c7912ef40c32d85f2e444b6b7c390e0eR11

[the remark this answers is not in the document]
Comment (simonihmig) [reply addon-test-support/-private/setup-window-mock.js]: Ah, you're right, of course. I was probably confusing this with `ember-mocha`, which supports the hooks-thing as well, to allow to use common `setupSomething()` functions to work with Mocha as well, and this only supports the *Each hooks: https://github.com/emberjs/ember-mocha/blob/master/addon-test-support/ember-mocha/setup-test.js#L70-L77

Any reason not to use `afterEach`?

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: > Ah, you're right, of course. I was probably confusing this with `ember-mocha`, which supports the hooks-thing as well, to allow to use common `setupSomething()` functions to work with Mocha as well, and this only supports the *Each hooks: https://github.com/emberjs/ember-mocha/blob/master/addon-test-support/ember-mocha/setup-test.js#L70-L77

I've opened https://github.com/emberjs/ember-mocha/issues/266 to track the addition of the `before` and `after` hooks, which actually are supported by `mocha`, but just not added to `ember-mocha`.

> Any reason not to use `afterEach`?

Purely for performance reasons. Since we already `reset` before each test, we don't need to `reset` after each test, but only after the last test, so that the next module can start with a clean slate.

Alternatively we could use `before` and `afterEach`, which would have the same effect.

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: Based on the discussion in https://github.com/emberjs/ember-mocha/issues/266#issuecomment-435835639, I would change this to _just_ `afterEach`, instead of `beforeEach` and `after`, if that's alright with you.

[the remark this answers is not in the document]
Comment (simonihmig) [reply addon-test-support/-private/setup-window-mock.js]: yes, makes sense. 

I am trying to think of any edge case where this could make a difference (i.e. would be a breaking change), but hopefully that's shouldn't be the case!?

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: This _could_ cause tests to fail, if you are mutating the `window` mock in a preceding module that _does not_ use `setupWindowMock`. Because then in the next module (that _does_ use `setupWindowMock`) the first test could fail, because the state would not have been reset until that very test has been run.

However, this is then indicative of a broken / leaky test in the first place.

Would there be an easy way to check whether the `window` mock is in pristine condition? If so, we could add an assertion in `before` to make sure that the previous module did clean up after itself.

**Comment by buschtoens** [comment]
@simonihmig Unless you know of a reasonably performant way to assert that the `window` mock is pristine, this PR is ready to be merged. ☺️

**Comment by simonihmig** [comment]
Released as 0.5.2!
